include/mbedtls/ssl.h

```diff
@@ -969,6 +957,60 @@ struct mbedtls_ssl_session
 #endif
 };
 
+/*
+ * Identifiers for PRFs used in various versions of TLS.
+ */
+typedef enum
+{
+   MBEDTLS_SSL_TLS_PRF_NONE,
+   MBEDTLS_SSL_TLS_PRF_SHA384,
+   MBEDTLS_SSL_TLS_PRF_SHA256,
+   MBEDTLS_SSL_HKDF_EXPAND_SHA384,
+   MBEDTLS_SSL_HKDF_EXPAND_SHA256
+}
+mbedtls_tls_prf_types;
+
+#if defined(MBEDTLS_SSL_EXPORT_KEYS)
```

Comment:
`MBEDTLS_SSL_EXPORT_KEYS` controls a very small amount of code, is not really a security risk (if you can corrupt the SSL context by adding a callback, you can surely corrupt the SSL context in many other ways), and has a wide variety of uses both for production and debugging. How about removing it (keeping it always on)? The fewer configuration options the better.

Reply:
I think this is probably a useful simplification, the impact of having it always-on looks extremely small.

Reply:
I disagree here and would like to keep the option. Key export is for two use cases: (a) debugging, (b) application-specific key export. Both are non-standard, and are another potential source of key leakage - users should thus have the ability to disable them. (One might even contemplate disabling them by default.) Finally, while I haven't measured it, multiple invocations from multiple places of an unknown function pointer, with multiple arguments each, should be enough code that you'd like to be able to remove it on a highly constrained system.

Long story short: No change here for now.

Reply:
> Both are non-standard, and are another potential source of key leakage - users should thus have the ability to disable them. (One might even contemplate disabling them by default.)

I don't see how that's relevant. The feature of invoking a callback is disabled by default, and even if you do register a callback, there is only a leak if the callback leaks the key.

> multiple invocations from multiple places of an unknown function pointer, with multiple arguments each

There's a single call site, that's not the issue. For an M0+ build of this PR at 95d69cf9471bd37cd5ea3553a17349e322eb69fc (baremetal config):
```
  350331  350027 (TOTALS)
   19971   19927 ssl_cli.o
   25364   25224 ssl_msg.o
   20787   20767 ssl_srv.o
   20246   20146 ssl_tls.o
```
For comparison, on `development` at 57864faf847572fa7b4e752068be55ce50bf2e8e:
```
  367608  367452 (TOTALS)
   20755   20751 ssl_srv.o
   20226   20102 ssl_tls.o
    4417    4389 version_features.o
```
I must say I'm surprised that the difference is so much larger in this PR — ~300B vs ~150B. I don't understand why there's an impact outside `ssl_tls.o`.

We're only testing the build with `MBEDTLS_SSL_EXPORT_KEYS` disabled in some reference configurations. So it's not untested, but it's slightly fragile (we may be missing some common combinations).

----

So now I would like to understand the code size impact, whether or not the compile-time option remains.

Reply:
> I don't see how that's relevant. The feature of invoking a callback is disabled by default, and even if you do register a callback, there is only a leak if the callback leaks the key.

If we never make mistakes, you're right. But the mere fact that we're checking for a callback and passing a key to it increases our attack surface.

Reply:
> So now I would like to understand the code size impact, whether or not the compile-time option remains.

I think we're making better use of our time by agreeing whether this feature can be disabled by default.

Reply:
The default build is the one meant to be used when linking the library dynamically. So there should be a strong reason to disable a feature. The security risk is barely perceptible (it's in TLS which uses plenty of other function pointers anyway), so not a reason to disable in dynamic builds.

Reply:
Fair point, if that's the established definition of the default config, I'm fine keeping it on by default.

It changes nothing w.r.t. my views to keep this configurable, though.

Reply:
Explanation for the code size difference: it's due to short vs long field accesses in the context. After moving the export fields near the end of the context, I get:
```
   20222   20146 ssl_tls.o
```
So:

1. Please move the new fields towards the end of the context. It's a preexisting issue that the order of fields doesn't put often-used fields near the beginning, but let's not make it worse.
2. I still think we should remove `MBEDTLS_SSL_EXPORT_KEYS`. However, since this wasn't in the original scope of this pull request, and it can be done in 3.x since we'd just be making the feature always on, this is not a blocker here.